### PR TITLE
Kotlinify FormatUtils

### PR DESCRIPTION
**Pull request by christopherniksch**
This PR is for the general goal of converting the project over to Kotlin

### CODE REVIEW

**Comment by cortinico** [comment]
Awesome! @redwarp would you be able to do this review?

**Comment by cortinico** [comment]
@christopherniksch You do have some detekt failures. Can I ask you to fix them?

```
Ruleset: exceptions - 40min debt
	TooGenericExceptionCaught - [e] at /home/travis/build/ChuckerTeam/chucker/library/src/main/java/com/chuckerteam/chucker/internal/support/FormatUtils.kt:48:18
	TooGenericExceptionCaught - [e] at /home/travis/build/ChuckerTeam/chucker/library/src/main/java/com/chuckerteam/chucker/internal/support/FormatUtils.kt:70:18
Ruleset: style - 40min debt
	NewLineAtEndOfFile - [FormatUtils.kt] at /home/travis/build/ChuckerTeam/chucker/library/src/main/java/com/chuckerteam/chucker/internal/support/FormatUtils.kt:160:1
	MaxLineLength - [<anonymous>] at /home/travis/build/ChuckerTeam/chucker/library/src/main/java/com/chuckerteam/chucker/internal/support/FormatUtils.kt:27:13
	MaxLineLength - [getShareText] at /home/travis/build/ChuckerTeam/chucker/library/src/main/java/com/chuckerteam/chucker/internal/support/FormatUtils.kt:83:9
	MaxLineLength - [<anonymous>] at /home/travis/build/ChuckerTeam/chucker/library/src/main/java/com/chuckerteam/chucker/internal/support/FormatUtils.kt:133:13
	MagicNumber - [unit] at /home/travis/build/ChuckerTeam/chucker/library/src/main/java/com/chuckerteam/chucker/internal/support/FormatUtils.kt:32:28
	MagicNumber - [unit] at /home/travis/build/ChuckerTeam/chucker/library/src/main/java/com/chuckerteam/chucker/internal/support/FormatUtils.kt:32:38
```

**Comment by christopherniksch** [comment]
@cortinico of course. I'll get a fix up later today

**Comment by redwarp** [comment]
Seems that there are still some formatting issues.

File: library/src/main/java/com/chuckerteam/chucker/internal/support/FormatUtils.kt [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+
+        return text
+    }
+
+    @JvmStatic
+    fun getShareCurlCommand(transaction: HttpTransaction): String {
+        var compressed = false
+        var curlCmd = "curl"
+        curlCmd += " -X " + transaction.method
+        val headers = transaction.getParsedRequestHeaders()
```
Comment (redwarp) [inline library/src/main/java/com/chuckerteam/chucker/internal/support/FormatUtils.kt]: Can be done on one line, `var curlCmd = "curl -X ${transaction.method}`

File: library/src/main/java/com/chuckerteam/chucker/internal/support/FormatUtils.kt [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+            serializer.transform(xmlSource, res)
+            ((res.outputStream as ByteArrayOutputStream).toByteArray()).toString()
+        } catch (e: TransformerException) {
+            xml
+        }
+    }
+
+    @JvmStatic
+    fun getShareText(context: Context, transaction: HttpTransaction): String {
+        var text = ""
```
Comment (redwarp) [inline library/src/main/java/com/chuckerteam/chucker/internal/support/FormatUtils.kt]: Could be cool if it was kotlinified even more and replaced raw string and the """ notation.

File: library/src/main/java/com/chuckerteam/chucker/internal/support/FormatUtils.kt [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+import kotlin.math.pow
+import org.xml.sax.InputSource
+
+internal object FormatUtils {
+
+    private const val SI_MULTIPLE = 1000
+    private const val BASE_TWO_MULTIPLE = 1024
+
+    fun formatHeaders(httpHeaders: List<HttpHeader>?, withMarkup: Boolean): String {
+        return httpHeaders?.joinToString(separator = "") { header ->
```
Comment (redwarp) [inline library/src/main/java/com/chuckerteam/chucker/internal/support/FormatUtils.kt]: A bit hard to read (to be fair, the java version as well), could you simply do a if else or when on markup and declare two raw string blocks?

File: library/src/main/java/com/chuckerteam/chucker/internal/support/FormatUtils.kt [hunk 4]
```diff
@@ -29,9 +29,11 @@ internal object FormatUtils {
 
     fun formatHeaders(httpHeaders: List<HttpHeader>?, withMarkup: Boolean): String {
         return httpHeaders?.joinToString(separator = "") { header ->
-            (if (withMarkup) "<b>" else "") + header.name + ": " +
-                (if (withMarkup) "</b>" else "") + header.value +
-                if (withMarkup) "<br />" else "\n"
+            if (withMarkup) {
+                "<b>" + header.name + ": " + "</b>" + header.value + "<br />"
+            } else {
+                header.name + ": " + header.value + "\n"
```
Comment (redwarp) [inline library/src/main/java/com/chuckerteam/chucker/internal/support/FormatUtils.kt]: Can you use when applicable the "${header.name}: ${header.value}\n" way of doing strings? Thanks

File: library/src/main/java/com/chuckerteam/chucker/internal/support/FormatUtils.kt [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+            serializer.transform(xmlSource, res)
+            ((res.outputStream as ByteArrayOutputStream).toByteArray()).toString()
+        } catch (e: TransformerException) {
+            xml
+        }
+    }
+
+    @JvmStatic
+    fun getShareText(context: Context, transaction: HttpTransaction): String {
+        var text = context.getString(R.string.chucker_url) + ": " + transaction.url + "\n"
```
Comment (cortinico) [inline library/src/main/java/com/chuckerteam/chucker/internal/support/FormatUtils.kt]: Can I ask you to use string interpolation here as well?

File: library/src/main/java/com/chuckerteam/chucker/internal/support/FormatUtils.kt [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+        text += "${context.getString(R.string.chucker_request_size)}: ${transaction.requestSizeString}\n"
+        text += "${context.getString(R.string.chucker_response_size)}: ${transaction.responseSizeString}\n"
+        text += "${context.getString(R.string.chucker_total_size)}: ${transaction.totalSizeString}\n"
+        text += "\n"
+        text += "---------- ${context.getString(R.string.chucker_request)} ----------\n\n"
+
+        var headers = formatHeaders(transaction.getParsedRequestHeaders(), false)
+
+        if (!TextUtils.isEmpty(headers)) {
+            text += headers + "\n"
```
Comment (cortinico) [inline library/src/main/java/com/chuckerteam/chucker/internal/support/FormatUtils.kt]: Can you use string interpolation here as well?

File: library/src/main/java/com/chuckerteam/chucker/internal/support/FormatUtils.kt [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+        }
+
+        text += if (transaction.isRequestBodyPlainText) {
+            transaction.getFormattedRequestBody()
+        } else {
+            context.getString(R.string.chucker_body_omitted)
+        }
+
+        text += "\n\n"
+        text += "---------- " + context.getString(R.string.chucker_response) + " ----------\n\n"
```
Comment (cortinico) [inline library/src/main/java/com/chuckerteam/chucker/internal/support/FormatUtils.kt]: Can you use string interpolation here as well?

File: library/src/main/java/com/chuckerteam/chucker/internal/support/FormatUtils.kt [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+
+        headers = formatHeaders(transaction.getParsedResponseHeaders(), false)
+
+        if (!TextUtils.isEmpty(headers)) {
+            text += headers + "\n"
```
Comment (cortinico) [inline library/src/main/java/com/chuckerteam/chucker/internal/support/FormatUtils.kt]: Can you use string interpolation here as well?

File: library/src/main/java/com/chuckerteam/chucker/internal/support/FormatUtils.kt (hunk 1 above)
Comment (cortinico) [reply library/src/main/java/com/chuckerteam/chucker/internal/support/FormatUtils.kt]: @redwarp comment was not addressed

File: library/src/main/java/com/chuckerteam/chucker/internal/support/FormatUtils.kt [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+            ) {
+                compressed = true
+            }
+            curlCmd += " -H \"$header.name: $header.value\""
+        }
+
+        val requestBody = transaction.requestBody
+        if (requestBody != null && requestBody.isNotEmpty()) {
+            // try to keep to a single line and use a subshell to preserve any line breaks
+            curlCmd += " --data $'" + requestBody.replace("\n", "\\n") + "'"
```
Comment (cortinico) [inline library/src/main/java/com/chuckerteam/chucker/internal/support/FormatUtils.kt]: Can you use string interpolation here as well?

**Comment by vbuberen** [comment]
Hey @christopherniksch.
It seems that this PR became quite old already. 
Do you have any plans on addressing those few comments left after reviews?
I could finish it myself, but want to be sure that I am not breaking any plans for you.

**Comment by vbuberen** [comment]
Addressed all comments, did a minor clean-up and merged latest `develop` changes.
@cortinico @redwarp Could you check it again, please?

File: library/src/main/java/com/chuckerteam/chucker/internal/support/FormatUtils.kt (hunk 1 above)
Comment (vbuberen) [reply library/src/main/java/com/chuckerteam/chucker/internal/support/FormatUtils.kt]: Addressed it.

**Comment by vbuberen** [comment]
Ok, I will merge it to not block overall development pace.
Thanks to this PR I was able to detect one general issue the library has at the moment. Going to open it to not lose track of it.

@christopherniksch Thanks a lot for your contribution.

**Pull request merged**